File: RCode/ESRL.R
ESRL<-function(x,rmin=.50,rmax=.85,B=1000){
sub_ESRL<-function(X0,rmin,rmax){
  n<-nrow(X0)
  K0 <- sample(2:(n/3 - 2), 1)
  r01 <- unique(sample(ncol(X0), replace = TRUE))
  X01 <- X0[, r01]
  s <- svd(X01)
  D <- diag(s$d)
  cu<-cumsum(s$d)/sum(s$d)
  bb<-runif(1,min=rmin, max=rmax)
  m0<-min(which(cu>bb))
  if (m0==1) {
    rec0<-s$u[,1:m0] * D[1:m0,1:m0] * (s$v[,1:m0])
  } else {
    rec0<-s$u[,1:m0] %*% D[1:m0,1:m0] %*% t(s$v[,1:m0])
  }
 return(kmeans(rec0,K0)$cluster)
}

hammingD<-DistM<-function(dat){
  Len<-dim(dat)
  ss<-Len[1]
  dismat<- matrix(1,ncol=ss,ss)
  for(i in 1:ss){
    if (i==ss) break
    for(j in (i:ss)){
      dismat[i,j]<-mean(dat[i,]==dat[j,],na.rm=T)
    }
  }
  return(1-t(dismat))
}


require("doParallel", character.only = T)
require("foreach", character.only = T)

cl <- makeCluster(detectCores()-1) 
registerDoParallel(cl) 
ens = foreach(i = 1:B,
              .combine = "rbind") %dopar% {
                fit1 <- sub_ESRL(x,knmin,knmax)
                fit1
              }
stopCluster(cl)



esrl<-hammingD(t(ens))
return(esrl)
}

File: PythonCode/ESRL.py

import numpy as np
import pandas as pd
from joblib import Parallel, delayed
import multiprocessing
from sklearn.cluster import KMeans


def sub_ESRL(X0,rmin,rmax):
n=X0.shape[0]
K0=np.random.choice(range(2,int(n/3)+1), size=1, replace=True, p=None)
r01=np.unique(np.random.choice(range(0,X0.shape[1]), size=X0.shape[1], replace=True))
X01=X0.iloc[:,r01]
#r01=r01-1 
U, s, VT =np.linalg.svd(X0,, full_matrices=False)
D=np.diag(s)
cu=s.cumsum()/sum(s)
bb= np.random.uniform(rmin,rmax,1)
m0=np.min(np.where(cu>bb))
if (m0==0):  m0=1

U=U[:,0:(np.shape(X0)[1]-1)]
rec0= U[:,:m0].dot(D[:m0,:m0].dot(VT[:m0,:m0]))
kmeans = KMeans(n_clusters=K0[0]).fit(rec0).labels_
    return(kmeans)

#def hammingD(dat):
#    ss=np.shape(dat)[1]
#    dismat=np.array([])
#    for i in range(ss-1):
#        for j in range(i+1,ss):
#        dismat[i,j]=np.mean(dat[:,i]==dat[:,j])
#    return(1-dismat)

def hammingD(dat):
    ss=np.shape(dat)[1]
    dismat=np.array([])
    for i in range(ss-1):
        for j in range(i+1,ss):
            dismat=np.append(dismat,np.mean(dat[:,i]==dat[:,j]))
    return(1-dismat)

def ESRL(x_dat,rmin=.50,rmax=.85,B=1000):
    num_cores = multiprocessing.cpu_count()
    par0=Parallel(n_jobs=num_cores-1)(
    delayed(sub_ESRL)(x_dat,rmin,rmax)
    for i in range(B))
    h=hammingD(np.transpose(par0).T)
    return(h)    

